### PR TITLE
[Micro:bit] Add micro:bit virtualBoard

### PR DESCRIPTION
**Pull request by fisher-alice**
This PR adds the ability for a user to run an App Lab project for which Maker Toolkit is enabled WITHOUT an actual micro:bit board connected. Note that [the PR to fix the VirtualBoard reset bug](https://github.com/code-dot-org/code-dot-org/pull/51101) was merged to this PR.

Currently, a user is able to run an App Lab project for which Maker Toolkit is enabled without an actual Circuit Playground board connected. However, if the project's Maker Toolkit is enabled with the micro:bit, an error results.

Here's a video of an App Lab project with Maker Toolkit enabled - first without a micro:bit connected, and then without a Circuit Playground connected. The program with the 'micro:bit' command `ledScreen.on` results in an error after the user selects 'Run Without Board'. But the program with the 'Circuit' command `led.on` runs so that the console statements are printed and an arc is drawn on the App screen.

https://user-images.githubusercontent.com/107423305/229821612-8e66c25b-9b12-4270-9f58-c6d5b94d808b.mp4

`FakeBoard` is the class used to allow Maker apps to run without an attached Circuit Playground. I updated this class to `VirtualCPBoard` and then added a `VirtualMBBoard`. These two classes attach fake, no-op components to the JS Interpreter. The classes are quite similar but the `VirtualCPBoard`'s `board` component is a `johnny-five` board controller while the `VirtualMBBoard`'s `board`component is a `MicroBitFirmataWrapper` which extends `MBFirmataClient` and is NOT an `EventEmitter`. I updated the name from 'fake' to 'virtual' [based on comments](https://github.com/code-dot-org/code-dot-org/pull/51068#discussion_r1157591046) from @jmkulwik and @breville that these classes should be architected as 'models' of virtual boards with the goal of eventually building an entire student-facing board simulator. 

I also decided to rename `MicrobitStubBoard` as `MBFirmataClientStub` because when I first encountered `MicrobitStubBoard`, I mistakenly thought it stubbed the `MicroBitBoard` and not the `boardClient_` in `MicroBitBoard.js`. The instance variable `boardClient_` is a `MicroBitFirmataWrapper` which extends `MBFirmataClient`. 

In `toolkit.js`, I added a check for which Maker API is set if `shouldRunWithVirtualBoard` returns `true` .

I refactored `MicroBitBoardTest.js` to export the function `itMakesMicroBitComponentsAvailable` so that it could be used in `VirtualBoardTest.js`.  And then `itImplementsTheMakerBoardInterface` set of tests were called on a `VirtualMBBoard` as well. I believe that there is now equitable test coverage for `VirtualMBBoard` and `VirtualCPBoard`. 

Here is a video of running an Maker Toolkit-enabled App Lab program - first without a micro:bit attached and then without a Circuit Playground attached:

 

https://user-images.githubusercontent.com/107423305/229839097-ee95c63b-ebf1-4e29-a8d9-632066dccec2.mp4

## Links
[jira - [Micro:bit] Allow user to run App Lab program without micro:bit board using FakeBoard](https://codedotorg.atlassian.net/browse/SL-582)

## Testing story
I added unit tests as explained above. 

## Follow-up work
- Refactor VirtualCPBoard and VirtualMBBoard and add user console statements in virtual function calls. Details in [jira ticket](https://codedotorg.atlassian.net/browse/SL-773).

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/lib/kits/maker/boards/FakeMBBoard.js [hunk 1]
```diff
@@ -0,0 +1,208 @@
+/** @file Fake for running Maker apps without an attached MicroBit board. */
```
Comment (jmkulwik) [inline apps/src/lib/kits/maker/boards/FakeMBBoard.js]: For a bit now, it's seemed like there's been a difference in expectations for the "Fake__Board" classes, where some people were thinking of it as an emulator and others were not. After looking at this file, I think the naming might be what's causing this. It struck me that this is more similar to a Mock that happens to get used in production. I wonder if there's a way we can indicate this in the naming or in the file header comments to resolve some confusion. Edit: After a little googling, I'm realizing Mock might too strongly indicate that this would be used for a test, but I still can't help but wonder if there's a better name for this (or maybe "Fake" is the correct name and I'm just not too familiar with hardware nomenclature).

Similarly if we do end up building an emulator, I wonder if we'd want it to hook into this or into something else.

Comment (breville) [reply apps/src/lib/kits/maker/boards/FakeMBBoard.js]: How about a "virtual board"?  

It does feel as though this should be architected as the "model" of a virtual board, and at some point we might be able to build the "view/controller" part.  Until then it's a "headless" virtual board, and if we get to build the other part, then we have an entire student-facing board simulator.

Comment (fisher-alice) [reply apps/src/lib/kits/maker/boards/FakeMBBoard.js]: Thanks to you both! I think 'VirtualCPBoard' and 'VirtualMBBoard' work well! @jmkulwik , lmk what you think.

Comment (jmkulwik) [reply apps/src/lib/kits/maker/boards/FakeMBBoard.js]: Works for me! If it's not too much to ask, ideally, components would follow the same convention. i.e. "VirtualMicroBitButton"

File: apps/src/lib/kits/maker/boards/FakeMBBoard.js [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+  }
+
+  /**
+   * @return {boolean} whether a real board is connected
+   */
+  boardConnected() {
+    return false;
+  }
+
+  createLed(pin) {
```
Comment (jmkulwik) [inline apps/src/lib/kits/maker/boards/FakeMBBoard.js]: Should these also have params, like the ones above?

File: apps/src/lib/kits/maker/boards/microBit/MBFirmataWrapper.js [hunk 3]
```diff
@@ -4,7 +4,7 @@ import {isWebSerialPort} from '../../util/boardUtils';
 
 export const ACCEL_EVENT_ID = 13;
 
-export default class MicrobitFirmataWrapper extends MBFirmataClient {
+export default class MicroBitFirmataWrapper extends MBFirmataClient {
```
Comment (jmkulwik) [inline apps/src/lib/kits/maker/boards/microBit/MBFirmataWrapper.js:7]: 🎉 Tada for consistency!!

File: apps/src/lib/kits/maker/toolkit.js [hunk 4]
```diff
@@ -32,7 +33,7 @@ import {getAppOptions} from '@cdo/apps/code-studio/initApp/loadApp';
 export {dropletConfig, configMicrobit, configCircuitPlayground, MakerError};
 
 /**
- * @type {CircuitPlaygroundBoard} The current board controller, populated when
+ * @type {CircuitPlaygroundBoard | FakeCPBoard || MicroBitBoard || FakeMBBoard} The current board controller, populated when
```
Comment (jmkulwik) [inline apps/src/lib/kits/maker/toolkit.js]: Super tiny NIT: should there be another vertical line between CircuitPlaygroundBoard | FakeCPBoard?

File: apps/src/lib/kits/maker/toolkit.js [hunk 5]
```diff
@@ -169,7 +170,12 @@ function getBoard() {
     makerBoardAPI === MB_API ? MicroBitBoard : CircuitPlaygroundBoard;
 
   if (shouldRunWithFakeBoard()) {
-    return Promise.resolve(new FakeBoard());
+    // Check which maker is being enabled
+    if (makerBoardAPI === MB_API) {
+      return Promise.resolve(new FakeMBBoard());
+    } else {
```
Comment (jmkulwik) [inline apps/src/lib/kits/maker/toolkit.js:183]: Do we need to account for a case where makerBoardAPI could be null or undefined?

File: apps/src/lib/kits/maker/boards/FakeMBBoard.js [hunk 6]
```diff
@@ -0,0 +1,208 @@
+/** @file Fake for running Maker apps without an attached MicroBit board. */
+import {EventEmitter} from 'events'; // provided by webpack's node-libs-browser
+import {MBFirmataClientStub} from '../../../../../test/unit/lib/kits/maker/boards/makeStubBoard';
```
Comment (jmkulwik) [inline apps/src/lib/kits/maker/boards/FakeMBBoard.js]: It feels a little odd to have test code in production for a couple reasons.
- It makes it a little harder to separate deploy packages from test packages (although I'd guess WebPack handles this gracefully for us, it's a bit of an odd pattern, so we may run into odd edge cases as we do version bumps in the future)
- When updating makeStubBoard in order to update a test, a maintainer wouldn't expect it to break production code since it's in a /test directory. 

A pattern I've seen work well in the past is to put these types of shared files in the production directory and have the tests import from there instead of the other way around. I've also seen a production util directory used for this as well.

File: apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js [hunk 7]
```diff
@@ -15,18 +15,18 @@ import ExternalLed from './ExternalLed';
  * Initializes a set of components for the currently
  * connected MicroBit board.
  *
- * @param {MBFirmataClient} board - Microbit firmata client
+ * @param {MBFirmataClient} boardClient - Microbit firmata client
```
Comment (jmkulwik) [inline apps/src/lib/kits/maker/boards/microBit/MicroBitComponents.js:18]: Nice renaming here!

File: apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js [hunk 8]
```diff
@@ -27,200 +27,216 @@ function stubOpenSerialPort(board) {
...
-  });
-
-  afterEach(() => {
-    board = undefined;
-    sinon.restore();
-  });
-
-  describe('Maker Board Interface', () => {
-    itImplementsTheMakerBoardInterface(MicroBitBoard, board => {
+export function itMakesMicroBitComponentsAvailable(
```
Comment (jmkulwik) [inline apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js:30]: I was surprised to see test code like this being exported. Can you share some details about why it made more sense to export instead of pulling the code from FakeBoardTest into here?

File: apps/src/lib/kits/maker/boards/FakeMBBoard.js (hunk 2 above)
Comment (jmkulwik) [reply apps/src/lib/kits/maker/boards/FakeMBBoard.js]: (sorry, I meant should they have doc comments regarding their params, like the ones above)

File: apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js [hunk 9]
```diff
@@ -27,200 +27,216 @@ function stubOpenSerialPort(board) {
...
 
-        it(`correct number of them`, () => {
-          expect(jsInterpreter.addCustomMarshalObject).to.have.callCount(
-            MB_COMPONENTS.length
+      MB_COMPONENTS.forEach(constructor => {
+        it(constructor, () => {
+          expect(jsInterpreter.globalProperties).to.have.ownProperty(
+            constructor
+          );
+          expect(jsInterpreter.globalProperties[constructor]).to.be.a(
```
Comment (jmkulwik) [inline apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js:84]: Big picture question: What risks would we see if we removed these tests?

Reasoning for asking: The main benefit I usually think of when it comes to unit tests is ensuring that branching conditions and business logic have protections from regressions. In this case, these tests seem to be checking whether each function exists, rather than what each function does. That makes me think this may be a section of code that doesn't need unit testing. Thoughts?

File: apps/src/lib/kits/maker/boards/FakeMBBoard.js (hunk 2 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/boards/FakeMBBoard.js]: Sure - and I'll add to `FakeCPBoard.js` as well.

File: apps/src/lib/kits/maker/toolkit.js (hunk 4 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/toolkit.js]: Thanks! I mean to be a single vertical line in-between the classes as it is [here](https://github.com/code-dot-org/code-dot-org/blob/9b40ce2be6aa137d72c14973f8acb717bae5c4d4/apps/src/lib/kits/maker/commands.js#L14).

File: apps/src/lib/kits/maker/toolkit.js (hunk 5 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/toolkit.js:183]: At this point `shouldRunWithFakeBoard()` returns `true` which means that `makerBoardAPI` will be assigned a value. The interesting thing is that if we are in a curriculum level with Circuit Playground enabled, `makerBoardAPI` is assigned the boolean `true` or the string `"circuitPlayground"` depending on the assignment from the `.level` file.  If we are in a standalone App Lab project, the `makerBoardAPI` is assigned the string `"circuitPlayground"`. The reason for 2 possible values for a CP is that when we only supported Circuit Playgrounds, the value of `makerlab_enabled` was either `true` or `false`. Once we began to support micro:bits, we changed the possible values to `"circuitPlayground"`, `"microbit"`, or `false`. 
For any level/standalone project with micro:bit enabled, the `makerBoardAPI` value is assigned `"microbit"`. 
Thus, I think that I prefer keeping as is. I consulted with @sanchitmalhotra126 about this and he agrees.

File: apps/src/lib/kits/maker/boards/FakeMBBoard.js (hunk 1 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/boards/FakeMBBoard.js]: Sure - that makes sense.

File: apps/src/lib/kits/maker/boards/FakeMBBoard.js (hunk 6 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/boards/FakeMBBoard.js]: This makes sense - thanks for the explanation. I'll go ahead and move `makeStubBoard.js` to src/lib/kits/maker/util.  While I'm here, I'll also rename the `makeStubBoard` function as `CPBoardStub`. Thanks!

File: apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js (hunk 8 above)
Comment (fisher-alice) [reply apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js:30]: Sure! The set of tests from `itImplementsTheMakerBoardInterface` is imported from `MakerBoardTest.js` and then called on for both `FakeCPBoard` and `FakeMBBoard` in `FakeBoardTest.js`.  The function `itMakesCircuitPlaygroundComponentsAvailable` imported from`CircuitPlaygroundBoardTest.js` is also called on in `FakeBoardTest.js`. Thus, I was following this pattern and exported `itMakesMicroBitComponentsAvailable` so that we can check that the `FakeMBBoard` also passes this set of tests.

File: apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js (hunk 9 above)
Comment (fisher-alice) [reply apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js:84]: These tests were helpful when I was writing `FakeMBBoard` - it ensured that I was including all of the components/constructors needed to create a 'virtual' version of the micro:bit board. It sounds like there is a possibility of moving towards a simulation of the boards so keeping these in place seems reasonable to me.

File: apps/src/lib/kits/maker/toolkit.js (hunk 5 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/toolkit.js:183]: Adding a bit more context: the function `getBoard` is called in the function `connect` also defined in `toolkit.js`. The [first check](https://github.com/code-dot-org/code-dot-org/blob/13a3d37c7e159d6ed597788b6f87b50a81116b96/apps/src/lib/kits/maker/toolkit.js#L67) in `connect` is ensuring that the Maker Toolkit is enabled. The Maker Toolkit [is enabled](https://github.com/code-dot-org/code-dot-org/blob/13a3d37c7e159d6ed597788b6f87b50a81116b96/apps/src/applab/applab.js#L697) in `applab.js` if `config.level.makerlabEnabled` is assigned a value. This same value is assigned to `makerBoardAPI` in the `getBoard` function.

File: apps/test/unit/lib/kits/maker/boards/microBit/AccelerometerTest.js [hunk 10]
```diff
@@ -1,6 +1,6 @@
 import {expect} from '../../../../../../util/reconfiguredChai';
 import Accelerometer from '@cdo/apps/lib/kits/maker/boards/microBit/Accelerometer';
-import {MicrobitStubBoard} from '../makeStubBoard';
+import {MBFirmataClientStub} from '../../../../../../../src/lib/kits/maker/util/makeStubBoard';
```
Comment (fisher-alice) [inline apps/test/unit/lib/kits/maker/boards/microBit/AccelerometerTest.js]: When I used the path '@cdo/lib/kits/maker/util/makeStubBoard', I received the following error:
Thus, I used this relative path instead. 
Screen Shot 2023-04-05 at 4 31 55 PM

File: apps/src/lib/kits/maker/toolkit.js (hunk 5 above)
Comment (jmkulwik) [reply apps/src/lib/kits/maker/toolkit.js:183]: 👍

File: apps/src/lib/kits/maker/toolkit.js (hunk 4 above)
Comment (jmkulwik) [reply apps/src/lib/kits/maker/toolkit.js]: 👍

File: apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js (hunk 9 above)
Comment (jmkulwik) [reply apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js:84]: 👍

File: apps/test/unit/lib/kits/maker/boards/microBit/AccelerometerTest.js (hunk 10 above)
Comment (jmkulwik) [reply apps/test/unit/lib/kits/maker/boards/microBit/AccelerometerTest.js]: I may be remembering wrong, but I think the issue is that it should be 
`@cdo/apps/lib/kits/maker/util/makeStubBoard`
rather than
`@cdo/lib/kits/maker/util/makeStubBoard`

Comment (fisher-alice) [reply apps/test/unit/lib/kits/maker/boards/microBit/AccelerometerTest.js]: Thanks! Updated paths.

File: apps/src/lib/kits/maker/boards/VirtualMBBoard.js [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+      const component = components[componentName];
+      jsInterpreter.createGlobalProperty(componentName, component);
+    }
+  }
+
+  /**
+   * @param {number} pin
+   * @param {number} modeConstant
+   */
+  pinMode(pin, modeConstant) {}
```
Comment (sanchitmalhotra126) [inline apps/src/lib/kits/maker/boards/VirtualMBBoard.js:88]: Probably a follow-up - would it make sense to extract some of these shared functions (pinMode, digitalWrite, etc) into a common base class (`BaseVirtualBoard` or something?)

File: apps/src/lib/kits/maker/toolkit.js (hunk 5 above)
Comment (sanchitmalhotra126) [reply apps/src/lib/kits/maker/toolkit.js:183]: Thanks for the details Alice! Might be good to just include a comment here that if shouldRunWithVirtualBoard() is true, then we can safely assume that MB_API should be set. If you'd like to include more context around the fact that MB_API may either be 'circuitPlayground' or true based on the value assigned in the level, then that might be helpful too, for posterity's sake

File: apps/src/lib/kits/maker/util/makeStubBoard.js [hunk 12]
```diff
@@ -1,7 +1,7 @@
 import five from '@code-dot-org/johnny-five';
 import Playground from 'playground-io';
 
-export function makeStubBoard() {
+export function CPBoardStub() {
```
Comment (sanchitmalhotra126) [inline apps/src/lib/kits/maker/util/makeStubBoard.js]: nit: might be good to retain the `make` verbiage as this is a function that returns a board (rather than an implementation of Board class)? maybe `makeCPBoardStub` or `makeStubCPBoard`?

**Review by sanchitmalhotra126 — APPROVED** [review]
LGTM!

File: apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js (hunk 8 above)
Comment (jmkulwik) [reply apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js:30]: Ok, there's something that's still feeling a little strange here, but I'll retain this as a non-blocking comment, especially since I can't put to words what's on my mind 😂

File: apps/src/lib/kits/maker/util/makeStubBoard.js (hunk 12 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/util/makeStubBoard.js]: Oh! That makes sense. Will update to `makeCPBoardStub`. Thanks!

File: apps/src/lib/kits/maker/boards/VirtualMBBoard.js (hunk 11 above)
Comment (fisher-alice) [reply apps/src/lib/kits/maker/boards/VirtualMBBoard.js:88]: Ooo - I like that idea. I also want to note that we talked about adding user console statements for the virtual board's function calls so the user can see that the calls were made successfully since we don't have a board simulator (yet).
I'll create a ticket to do both of these things.

Comment (fisher-alice) [reply apps/src/lib/kits/maker/boards/VirtualMBBoard.js:88]: https://codedotorg.atlassian.net/browse/SL-773

File: apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js (hunk 8 above)
Comment (fisher-alice) [reply apps/test/unit/lib/kits/maker/boards/microBit/MicroBitBoardTest.js:30]: I'm intrigued about why this feels a little strange to you and definitely open to discussion. I'll add a note to discuss during our next 1-1 mtg and am also open to refactoring these tests during the time set aside this summer for 'Maker Cleanup'.
